src/select.test.ts

```diff
@@ -0,0 +1,222 @@
+        { bar: boolean } | {}
+      >();
+    });
+
+    it("converts primitives to empty object", () => {
+      const standaloneSelect = q.__experimental_select({
+        "foo > 2": ["bar", z.boolean()],
+        default: {
+          baz: z.string(),
+        },
+      });
+      const entityQuery = q("foo").__experimental_select(standaloneSelect);
+
+      expect(entityQuery.query).toBe("foo{...select(foo > 2 => bar, { baz })}");
+      expectTypeOf({} as z.infer<typeof entityQuery.schema>).toEqualTypeOf<
+        {} | { baz: string }
+      >();
+    });
+
+    it("handles nested union types properly", () => {
+      const nestedSelect = q.__experimental_select({
+        "foo == 3": ["a", q.string()],
+        default: {
+          b: q.boolean(),
+        },
+      });
+
+      const standaloneSelect = q.__experimental_select({
+        "foo > 2": nestedSelect,
+        default: {
+          baz: z.string(),
+        },
+      });
+      const entityQuery = q("foo").__experimental_select(standaloneSelect);
+
+      expect(entityQuery.query).toBe(
+        "foo{...select(foo > 2 => select(foo == 3 => a, { b }), { baz })}"
+      );
+      expectTypeOf({} as z.infer<typeof entityQuery.schema>).toEqualTypeOf<
+        {} | { b: boolean } | { baz: string }
+      >();
+    });
+  });
+
+  describe("with {condition: Selection} input", () => {
+    it("updates query and handles nested object types", () => {
+      const entityQuery = q("foo").__experimental_select({
+        "foo > 2": {
+          bar: z.boolean(),
+        },
+        default: {
+          baz: z.string(),
+        },
+      });
+
+      expect(entityQuery.query).toBe(
+        "foo{...select(foo > 2 => { bar }, { baz })}"
+      );
+      expectTypeOf({} as z.infer<typeof entityQuery.schema>).toEqualTypeOf<
+        { bar: boolean } | { baz: string }
+      >();
+    });
+  });
+});
+
+describe("ArrayQuery.select()", () => {
+  it("handles standalone select input", () => {
+    const standaloneSelect = q.__experimental_select({
+      "foo > 2": {
+        bar: z.boolean(),
+      },
+      default: {
+        baz: z.string(),
+      },
+    });
+    const query = q("*").filter().__experimental_select(standaloneSelect);
+
+    expect(query.query).toBe("*[]{...select(foo > 2 => { bar }, { baz })}");
+    expectTypeOf({} as z.infer<typeof query.schema>).toEqualTypeOf<
+      ({ bar: boolean } | { baz: string })[]
+    >();
+  });
+
+  it("handles {condition: Selection} input", () => {
+    const query = q("*")
+      .filter()
+      .__experimental_select({
+        "foo > 2": {
+          bar: z.boolean(),
+        },
+        default: {
+          baz: z.string(),
+        },
+      });
+
+    expect(query.query).toBe("*[]{...select(foo > 2 => { bar }, { baz })}");
+    expectTypeOf({} as z.infer<typeof query.schema>).toEqualTypeOf<
+      ({ bar: boolean } | { baz: string })[]
+    >();
+  });
+});
```

Comment:
Can you add a test case or two for the non-spreading projection scenario (`#3` in your PR notes)?

Reply:
good catch. yup, will do

Reply:
added in the latest commits